L96-python-setuptools-plugin.md

```diff
@@ -0,0 +1,107 @@
+Setuptools entrypoint for generating file
+----
+* Author(s): Patrick Lannigan
+* Approver: a11r
+* Status: Draft
+* Implemented in: Python
+* Last updated: 2022-02-18
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Implement a `setuptools` entrypoint for generating files when a Python distribution is being built.
+This type of integration would be called automatically by `setuptools` and would not require and
+explicit call by the project to generate the files. The entrypoint would be configured using an
+established python project configuration file and would only require the `grpc-tools` package to be
+listed as a build time dependency.
+
+## Background
+
+Currently `grpc_tools` [provides a command][grpc_command] for integrating with `setuptools`. However,
+this requires custom integration on the part of the package (example:
+[grpc_health_checking][grpc_command_example_integration] and
+[explicitly calling the command][grpc_command_example_call] when building the wheel).
+
+With the introduction of [PEP 517][pep_517] and [PEP 518][pep_518], directly calling `setup.py` will
+start becoming less common. Additionally, the use of custom commands will not work with projects that
+use this mechanism and solely rely on [setup.cfg][declarative_config].
+
+Custom commands are not the only integration point that is available. `setuptools` also supports
+entrypoints that allows for packages to act as a plugin and provide custom behavior. This is
+[documented with examples][setuptools_entrypoint] of generating files based on revision control
+information. This could be used to generate source files based on `.proto` files.
+
+[Original Feature Request][original_feature_request]
+
+## Proposal
+
+Add an entrypoint to `grpc-tools` that would register as a `file_finer` hook with `setuptools`. To
+simplify the implementation without introducing an additional code path, the entrypoint will target
+the same `main()` function used by the `grpc_tools.protoc` CLI. A [tool table][tool_table] within the
+file `pyproject.toml` will be used to accept configuration parameters. This will add a dependency on
+tomli for now, but this will [not be needed][pep_680] starting in Python 3.11.
+
```

Comment:
Optional: The import path generated by `grpc_tools.protoc` can be counterintuitive https://github.com/grpc/grpc/issues/9575. In the config file, we might want to provide the flexibility to: 1. set the work directory; 2. the ["prefixes_to_filter"](https://github.com/grpc/grpc/blob/2d4f3c56001cd1e1f85734b2f7c5ce5f2797c38a/src/compiler/python_generator.h#L39) argument.

CC @gnossen , if you got a second, can you help to chill in about how to use "prefixes_to_filter" in our generator?

Reply:
`prefixes_to_filter` is currently not available in `grpcio-tools`, only in the Bazel rules, so I think it's okay to leave it out of this proposal, at least initially. Regardless, [this](https://github.com/grpc/grpc/pull/20141) is the best reference on usage.